### ISSUE TITLE
[HELP-WANTED] macOS Support improvements and testing

### ISSUE THREAD

**Issue by mazen160** [ProtonVPN/protonvpn-cli#65]
Hi all,

First of all, thank you everyone for your help in protonvpn-cli. It has been a young journey, but the progress we did is very impressive.

The next milestone we have is to confirm the macOS support for protonvpn-cli. The problems we have:

* Issues in DNS management.
There is a built-in function that does the DNS management in order to protect against DNS leaks.

We need to add support for macOS in this part.

* Issues in IPv6.
There is a built-in function that does the ipv6 handling. This protects against original IP leaks via ipv6 addresses.

We need to add support to handle ipv6 correctly in macOS.

If you are a macOS app user/developer, please help us by submitting a PR with fixes.

Also, the PR needs to be tested well, we will post PRs here for testers to confirm if it adds support on their machines.

Thanks again!
Mazin

**Comment by mazen160** [comment]
A good place to start is to check this commit: https://github.com/ProtonVPN/protonvpn-cli/commit/717f0ef28d437ac6747572f0176a0a3df1431d4f

These are simple preparations that can help guiding you where to start. Also, to implement the addition without breaking support for other distros.

**Comment by xilopaint** [comment]
I think the first thing to decide regarding the DNS issue before start writing the code is whether we will use `scutil` (adopted for Tunnelblick) or `networksetup` (seemingly simpler).

I guess `scutil` is more recommended but I confess that I don't know its pros over `networksetup`.

A summary of the issue can be found in #51.

**Comment by xilopaint** [comment]
@mazen160, can you check with the team of the macOS app which command they used to handle the DNS update? I'm almost certain it must have been one of the two mentioned.

**Comment by xilopaint** [comment]
#66 adds IPv6 management for macOS.

**Comment by rob-patchett** [comment]
Hey @xilopaint, Rob here from the macOS app team. We are using Apple's Network Extension framework along with IKEv2, so the DNS server configuration is provided by the server and the app itself doesn't require explicit DNS management.

Unfortunately I don't have much experience with this kind of issue, so can't really comment on which route to take.

**Comment by xilopaint** [comment]
Nice, Rob! Do you use Swift or Objective-C?

Btw, could you test #66 implementation for IPv6 management and tell me if it works for you?

**Comment by rob-patchett** [comment]
Yeah, the app is entirely in Swift (though I have strong experience in both).

Will test out your PR and get back to ya.

**Comment by rob-patchett** [comment]
@xilopaint, looks good to me 🙂
<img width="1183" alt="screen shot 2018-04-02 at 11 00 39" src="https://user-images.githubusercontent.com/5439648/38190668-911d510c-3665-11e8-9be2-bd87a37571a8.png">

**Comment by Rafficer** [comment]
Btw, if any of you need more testers, I'm sure reddit can help :P

**Comment by xilopaint** [comment]
I think it's enough. Let's wait @mazen160 to review the PR.

**Comment by mazen160** [comment]
Awesome!

We now only have the DNS issue to be managed.

**Comment by mazen160** [comment]
I added simple addition the PR, ipv6 handling should be supported now on macOS

**Comment by cantab412** [comment]
@xilopaint - Hi, just getting back from Easter holiday.

Are DNS leaks a known issue?  If not, I'll open a new one:  https://dnsleaktest.com/ 

Tried updating/running and while it works, it is still complaining about IPv6 mangement.  I do not have IPv6 enabled on my router, and on my Mac it is set to "Automatic" (see below).

```
$ sudo ./protonvpn-cli.sh -update
[#] Checking for update.
[*] protonvpn-cli is up-to-date!

$ sudo ./protonvpn-cli.sh -r
Fetching ProtonVPN Servers...
Connecting...
[$] Connected!
[#] New IP: 89.39.107.209

$ sudo ./protonvpn-cli.sh -d
Disconnecting...
cat: /Users/cantab412/.protonvpn-cli/.ipv6_address: No such file or directory
./protonvpn-cli.sh: line 183: ip: command not found
[!] There are issues in managing ipv6 in the system. Please test the system for the root cause.
Not able to manage ipv6 by protonvpn-cli might cause issues in leaking the system's ipv6 address.
[#] Disconnected.
[#] Current IP: [redacted]
```

While connected to VPN, checking to see if .ipv6_address exists (it does not):

```
$ sudo ls -la .protonvpn-cli/

total 16
dr--------   4 root  wheel   136 29 Mar 06:49 .
drwxr-xr-x+ 61 cant  staff  2074  2 Apr 14:24 ..
-r--------   1 root  wheel    50 29 Mar 06:49 protonvpn_openvpn_credentials
-r--------   1 root  wheel     2 29 Mar 06:49 protonvpn_tier
```

Here are some ifconfig results when connected/not connected to the VPN:

[ifconfig-connected.txt](https://github.com/ProtonVPN/protonvpn-cli/files/1868168/ifconfig-connected.txt)
[ifconfig-not-connected.txt](https://github.com/ProtonVPN/protonvpn-cli/files/1868169/ifconfig-not-connected.txt)

![screen shot 2018-04-02 at 14 41 18](https://user-images.githubusercontent.com/25002291/38198352-52a2355e-3684-11e8-9dc4-39b3519976fe.png)

**Comment by cantab412** [comment]
...ah.  It would appear that my ISP does not currently support IPv6, so I am probabaly not a good person to be testing this.  

Could be a good check to add, though?

**Comment by mazen160** [comment]
@cantab412 

Yes it's a known issue on maOS that requires a fix. We should be fixing it on this ticket.

If you encounter any DNS leaks on Linux, please let us know

**Comment by xilopaint** [comment]
> Tried updating/running and while it works, it is still complaining about IPv6 mangement. I do not have IPv6 enabled on my router, and on my Mac it is set to "Automatic" (see below).
> 
> $ sudo ./protonvpn-cli.sh -update
> [#] Checking for update.
> [*] protonvpn-cli is up-to-date!
> 
> $ sudo ./protonvpn-cli.sh -r
> Fetching ProtonVPN Servers...
> Connecting...
> [$] Connected!
> [#] New IP: 89.39.107.209
> 
> $ sudo ./protonvpn-cli.sh -d
> Disconnecting...
> cat: /Users/cantab412/.protonvpn-cli/.ipv6_address: No such file or directory
> ./protonvpn-cli.sh: line 183: ip: command not found
> [!] There are issues in managing ipv6 in the system. Please test the system for the root cause.
> Not able to manage ipv6 by protonvpn-cli might cause issues in leaking the system's ipv6 address.
> [#] Disconnected.
> [#] Current IP: [redacted]

@cantab412, why are you entering `./protonvpn-cli.sh` instead of `protonvpn-cli`? It should be the cause you're still seeing the IPv6 error.

**Comment by cantab412** [comment]
@xilopaint Hmm - didn't notice the syntax had changed, or that I was calling it incorrectly the whole time (which I wouldn't doubt)


```
$ sudo protonvpn-cli -f
Fetching ProtonVPN Servers...
Connecting...
[$] Connected!
[#] New IP: 62.112.9.167

$ sudo protonvpn-cli -d
Disconnecting...
[#] Disconnected.
[#] Current IP: [redacted]
$
```

**Comment by mazen160** [comment]
@xilopaint Thanks for the PR. It would be merged along with the final fix for the improvement 👍 
PR: https://github.com/ProtonVPN/protonvpn-cli/pull/72

**Comment by rob-patchett** [comment]
Sorry for the long wait folk. Please take a look at this PR for DNS management on Mac: #95 (thanks @xilopaint, I pulled in your work from #72).

Can someone please test this out on Linux in addition to anyone who can test on macOS. Thanks!

**Comment by xilopaint** [comment]
> Can someone please test this out on Linux in addition to anyone who can test on macOS. Thanks!

It seems I have #73 on macOS.

**Comment by rob-patchett** [comment]
By the looks of it, the issue should be identical on all platforms. It seems this is still an issue in general for some servers. @xilopaint are you able to test master on Linux and confirm if this is a regression or not?

**Comment by xilopaint** [comment]
@rob-patchett yes, I can confirm the same issue in Linux.

@mazen160 should #73 be reopen?

**Comment by mazen160** [comment]
Re: DNS management issue of free plans --> paid servers is now re-opened.There seems to be a bug somewhere.

Re: PR, I will be merging it next.
